File: python/schemas/categories/component.py
"""Component tool schema definitions for KiCAD MCP Server.

Following the MCP 2025-06-18 specification for tool definitions.
Split out of the former monolithic schemas/tool_schemas.py.
"""

COMPONENT_TOOLS = [
    {
        "name": "place_component",
        "title": "Place Component",
        "description": "Places a component with specified footprint at given coordinates on the board.",
        "inputSchema": {
            "type": "object",
            "properties": {
                "reference": {
                    "type": "string",
                    "description": "Component reference designator (e.g., R1, C2, U3)",
                },
                "footprint": {
                    "type": "string",
                    "description": "Footprint library:name (e.g., Resistor_SMD:R_0805_2012Metric)",
                },
                "x": {"type": "number", "description": "X coordinate in millimeters"},
                "y": {"type": "number", "description": "Y coordinate in millimeters"},
                "rotation": {
                    "type": "number",
                    "description": "Rotation angle in degrees (0-360)",
                    "minimum": 0,
                    "maximum": 360,
                    "default": 0,
                },
                "layer": {
                    "type": "string",
                    "enum": ["F.Cu", "B.Cu"],
                    "description": "Board layer (top or bottom)",
                    "default": "F.Cu",
                },
            },
            "required": ["reference", "footprint", "x", "y"],
        },
    },
    {
        "name": "move_component",
        "title": "Move Component",
        "description": "Moves an existing component to a new position on the board.",
        "inputSchema": {
            "type": "object",
            "properties": {
                "reference": {
                    "type": "string",
                    "description": "Component reference designator",
                },
                "x": {
                    "type": "number",
                    "description": "New X coordinate in millimeters",
                },
                "y": {
                    "type": "number",
                    "description": "New Y coordinate in millimeters",
                },
            },
            "required": ["reference", "x", "y"],
        },
    },
    {
        "name": "rotate_component",
        "title": "Rotate Component",
        "description": "Rotates a component by specified angle. Rotation is cumulative with existing rotation.",
        "inputSchema": {
            "type": "object",
            "properties": {
                "reference": {
                    "type": "string",
                    "description": "Component reference designator",
                },
                "angle": {
                    "type": "number",
                    "description": "Rotation angle in degrees (positive = counterclockwise)",
                },
            },
            "required": ["reference", "angle"],
        },
    },
    {
        "name": "delete_component",
        "title": "Delete Component",
        "description": "Removes a component from the board.",
        "inputSchema": {
            "type": "object",
            "properties": {
                "reference": {
                    "type": "string",
                    "description": "Component reference designator",
                }
            },
            "required": ["reference"],
        },
    },
    {
        "name": "edit_component",
        "title": "Edit Component Properties",
        "description": "Modifies properties of an existing component (value, footprint, etc.).",
        "inputSchema": {
            "type": "object",
            "properties": {
                "reference": {
                    "type": "string",
                    "description": "Component reference designator",
                },
                "value": {"type": "string", "description": "New component value"},
                "footprint": {
                    "type": "string",
                    "description": "New footprint library:name",
                },
            },
            "required": ["reference"],
        },
    },
    {
        "name": "get_component_properties",
        "title": "Get Component Properties",
        "description": "Retrieves detailed properties of a specific component.",
        "inputSchema": {
            "type": "object",
            "properties": {
                "reference": {
                    "type": "string",
                    "description": "Component reference designator",
                }
            },
            "required": ["reference"],
        },
    },
    {
        "name": "get_component_list",
        "title": "List All Components",
        "description": "Returns a list of all components on the board with their properties.",
        "inputSchema": {"type": "object", "properties": {}},
    },
    {
        "name": "find_component",
        "title": "Find Components",
        "description": "Searches for components matching specified criteria. Supports partial matching on reference, value, or footprint patterns.",
        "inputSchema": {
            "type": "object",
            "properties": {
                "reference": {
                    "type": "string",
                    "description": "Reference designator pattern to match (e.g., 'R1', 'U', 'C2')",
                },
                "value": {
                    "type": "string",
                    "description": "Value pattern to match (e.g., '10k', '100nF')",
                },
                "footprint": {
                    "type": "string",
                    "description": "Footprint pattern to match (e.g., '0805', 'SOIC')",
                },
            },
        },
    },
    {
        "name": "get_component_pads",
        "title": "Get Component Pads",
        "description": "Returns all pads for a component with their positions, net connections, sizes, and shapes.",
        "inputSchema": {
            "type": "object",
            "properties": {
                "reference": {
                    "type": "string",
                    "description": "Component reference designator (e.g., U1, R5)",
                }
            },
            "required": ["reference"],
        },
    },
    {
        "name": "get_pad_position",
        "title": "Get Pad Position",
        "description": "Returns the position and properties of a specific pad on a component.",
        "inputSchema": {
            "type": "object",
            "properties": {
                "reference": {
                    "type": "string",
                    "description": "Component reference designator",
                },
                "padName": {
                    "type": "string",
                    "description": "Pad name or number (e.g., '1', '2', 'A1')",
                },
                "padNumber": {
                    "type": "string",
                    "description": "Alternative to padName - pad number",
                },
            },
            "required": ["reference"],
        },
    },
    {
        "name": "place_component_array",
        "title": "Place Component Array",
        "description": "Places multiple copies of a component in a grid or circular pattern.",
        "inputSchema": {
            "type": "object",
            "properties": {
                "referencePrefix": {
                    "type": "string",
                    "description": "Reference prefix (e.g., 'R' for R1, R2, R3...)",
                },
                "startNumber": {
                    "type": "integer",
                    "description": "Starting number for references",
                    "minimum": 1,
                    "default": 1,
                },
                "footprint": {
                    "type": "string",
                    "description": "Footprint library:name",
                },
                "pattern": {
                    "type": "string",
                    "enum": ["grid", "circular"],
                    "description": "Array pattern type",
                },
                "count": {
                    "type": "integer",
                    "description": "Total number of components to place",
                    "minimum": 1,
                },
                "startX": {
                    "type": "number",
                    "description": "Starting X coordinate in millimeters",
                },
                "startY": {
                    "type": "number",
                    "description": "Starting Y coordinate in millimeters",
                },
                "spacingX": {
                    "type": "number",
                    "description": "Horizontal spacing in mm (for grid pattern)",
                },
                "spacingY": {
                    "type": "number",
                    "description": "Vertical spacing in mm (for grid pattern)",
                },
                "radius": {
                    "type": "number",
                    "description": "Circle radius in mm (for circular pattern)",
                },
                "rows": {
                    "type": "integer",
                    "description": "Number of rows (for grid pattern)",
                    "minimum": 1,
                },
                "columns": {
                    "type": "integer",
                    "description": "Number of columns (for grid pattern)",
                    "minimum": 1,
                },
            },
            "required": [
                "referencePrefix",
                "footprint",
                "pattern",
                "count",
                "startX",
                "startY",
            ],
        },
    },
    {
        "name": "align_components",
        "title": "Align Components",
        "description": "Aligns multiple components horizontally or vertically.",
        "inputSchema": {
            "type": "object",
            "properties": {
                "references": {
                    "type": "array",
                    "description": "Array of component reference designators to align",
                    "items": {"type": "string"},
                    "minItems": 2,
                },
                "direction": {
                    "type": "string",
                    "enum": ["horizontal", "vertical"],
                    "description": "Alignment direction",
                },
                "spacing": {
                    "type": "number",
                    "description": "Spacing between components in mm (optional, for even distribution)",
                },
            },
            "required": ["references", "direction"],
        },
    },
    {
        "name": "check_courtyard_overlaps",
        "title": "Check Courtyard Overlaps",
        "description": (
            "Detects courtyard overlaps between footprints and (optionally) flags "
            "footprints whose courtyard extends past the board outline. "
            "Returns overlap pairs with intersection extents and per-component "
            "boundary violations, both in mm. Accepts a 'positions' dict to "
            "evaluate a HYPOTHETICAL placement without modifying the board — "
            "use this before committing a move_component / place_component call "
            "to know if it will trigger DRC. "
            "Approach ported from morningfire-pcb-automation "
            "(https://github.com/NiNjA-CodE/morningfire-pcb-automation, "
            "scripts/placement/check_overlaps.py); this version reads real "
            "courtyard polygons from the board (not a static lookup table) and "
            "supports virtual placement + rotation + clearance margin."
        ),
        "inputSchema": {
            "type": "object",
            "properties": {
                "positions": {
                    "type": "object",
                    "description": (
                        "Virtual placements: map of reference designator to "
                        "[x, y] or [x, y, rotation_degrees] in mm. Each listed "
                        "ref is checked AS IF it were at the given coordinates. "
                        "Unspecified refs use their current board position."
                    ),
                    "additionalProperties": {
                        "type": "array",
                        "items": {"type": "number"},
                        "minItems": 2,
                        "maxItems": 3,
                    },
                },
                "refs": {
                    "type": "array",
                    "description": (
                        "Limit the check to these refs (default: every " "footprint on the board)."
                    ),
                    "items": {"type": "string"},
                },
                "margin": {
                    "type": "number",
                    "description": (
                        "Extra clearance in mm added around every courtyard "
                        "(default 0). Useful to enforce a manufacturing keepout "
                        "wider than the symbol's declared courtyard."
                    ),
                    "default": 0,
                },
                "include_boundary": {
                    "type": "boolean",
                    "description": (
                        "Also flag courtyards that extend past the board outline " "(default true)."
                    ),
                    "default": True,
                },
                "board_outline": {
                    "type": "object",
                    "description": (
                        "Optional override for the board outline bbox. Default: "
                        "derived from Edge.Cuts."
                    ),
                    "properties": {
                        "x1": {"type": "number"},
                        "y1": {"type": "number"},
                        "x2": {"type": "number"},
                        "y2": {"type": "number"},
                        "unit": {
                            "type": "string",
                            "enum": ["mm", "mil", "inch"],
                            "default": "mm",
                        },
                    },
                    "required": ["x1", "y1", "x2", "y2"],
                },
            },
        },
    },
    {
        "name": "duplicate_component",
        "title": "Duplicate Component",
        "description": "Creates a copy of an existing component with new reference designator.",
        "inputSchema": {
            "type": "object",
            "properties": {
                "sourceReference": {
                    "type": "string",
                    "description": "Reference of component to duplicate",
                },
                "newReference": {
                    "type": "string",
                    "description": "Reference designator for the new component",
                },
                "offsetX": {
                    "type": "number",
                    "description": "X offset from original position in mm",
                    "default": 0,
                },
                "offsetY": {
                    "type": "number",
                    "description": "Y offset from original position in mm",
                    "default": 0,
                },
            },
            "required": ["sourceReference", "newReference"],
        },
    },
]

File: src/tools/component.ts
/**
 * Component management tools for KiCAD MCP server
 */

import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";
import { logger } from "../logger.js";
import { boundingBoxFilter, paginationParams } from "./pagination-params.js";
import { CommandFunction, formatKicadResult, makePassthrough } from "./tool-response.js";

/**
 * Register component management tools with the MCP server
 *
 * @param server MCP server instance
 * @param callKicadScript Function to call KiCAD script commands
 */
export function registerComponentTools(server: McpServer, callKicadScript: CommandFunction): void {
  const passthrough = makePassthrough(callKicadScript);
  logger.info("Registering component management tools");

  const placeComponentSchema = {
    componentId: z
      .string()
      .describe("Footprint library ID, e.g. 'Package_DIP:DIP-8_W7.62mm' or 'R_0603_10k'."),
    position: z
      .object({
        x: z.number(),
        y: z.number(),
        unit: z.enum(["mm", "inch", "mil"]).optional().describe("Unit (default mm)"),
      })
      .describe("Position coordinates (unit defaults to mm)"),
    reference: z.string().optional().describe("Optional desired reference (e.g., 'R5')"),
    value: z.string().optional().describe("Optional component value (e.g., '10k')"),
    footprint: z.string().optional().describe("Optional specific footprint name"),
    rotation: z.number().optional().describe("Optional rotation in degrees"),
    layer: z.string().optional().describe("Optional layer (e.g., 'F.Cu', 'B.SilkS')"),
    boardPath: z
      .string()
      .optional()
      .describe("Path to the .kicad_pcb — required for project-local footprint libraries"),
  } as const;

  type PlaceComponentArgs = {
    componentId: string;
    position: { x: number; y: number; unit?: "mm" | "inch" | "mil" };
    reference?: string;
    value?: string;
    footprint?: string;
    rotation?: number;
    layer?: string;
    boardPath?: string;
  };

  const placeComponentHandler = async (args: PlaceComponentArgs) => {
    logger.debug(
      `Placing component: ${args.componentId} at ${args.position.x},${args.position.y} ${args.position.unit}`,
    );
    const result = await callKicadScript("place_component", args);
    return formatKicadResult(result);
  };

  server.tool(
    "place_component",
    "Add a NEW footprint instance to the PCB. Errors if the reference already exists — use move_component to relocate an existing part.",
    placeComponentSchema,
    placeComponentHandler,
  );

  server.tool(
    "move_component",
    "Move a PCB component to a new position. Optionally update rotation or flip to a different copper layer.",
    {
      reference: z.string().describe("Reference designator of the component (e.g., 'R5')"),
      position: z
        .object({
          x: z.number(),
          y: z.number(),
          unit: z.enum(["mm", "inch", "mil"]).optional().describe("Unit (default mm)"),
        })
        .describe("New position coordinates (unit defaults to mm)"),
      rotation: z.number().optional().describe("Optional new rotation in degrees"),
      layer: z
        .string()
        .optional()
        .describe("Optional target layer (e.g., 'F.Cu', 'B.Cu') - flips component if needed"),
      allowOffBoard: z
        .boolean()
        .optional()
        .describe(
          "Allow moving the component outside the board outline. Off-board targets are refused by default (errorCode POSITION_OFF_BOARD); set true to place a part off the board intentionally.",
        ),
    },
    async ({ reference, position, rotation, layer, allowOffBoard }) => {
      logger.debug(
        `Moving component: ${reference} to ${position.x},${position.y} ${position.unit}`,
      );
      const result = await callKicadScript("move_component", {
        reference,
        position,
        rotation,
        layer,
        allowOffBoard,
      });

      return formatKicadResult(result);
    },
  );

  server.tool(
    "rotate_component",
    "Rotate a PCB component to an absolute angle in degrees.",
    {
      reference: z.string().describe("Reference designator of the component (e.g., 'R5')"),
      angle: z.number().describe("Rotation angle in degrees (absolute, not relative)"),
    },
    async ({ reference, angle }) => {
      logger.debug(`Rotating component: ${reference} to ${angle} degrees`);
      const result = await callKicadScript("rotate_component", {
        reference,
        angle,
      });

      return formatKicadResult(result);
    },
  );

  server.tool(
    "delete_component",
    "Remove a component from the PCB by its reference designator.",
    {
      reference: z
        .string()
        .describe("Reference designator of the component to delete (e.g., 'R5')"),
    },
    async ({ reference }) => {
      logger.debug(`Deleting component: ${reference}`);
      const result = await callKicadScript("delete_component", { reference });

      return formatKicadResult(result);
    },
  );

  server.tool(
    "edit_component",
    "Edit properties of an existing PCB component (reference, value, footprint).",
    {
      reference: z.string().describe("Reference designator of the component (e.g., 'R5')"),
      newReference: z.string().optional().describe("Optional new reference designator"),
      value: z.string().optional().describe("Optional new component value"),
      footprint: z.string().optional().describe("Optional new footprint"),
    },
    async ({ reference, newReference, value, footprint }) => {
      logger.debug(`Editing component: ${reference}`);
      const result = await callKicadScript("edit_component", {
        reference,
        newReference,
        value,
        footprint,
      });

      return formatKicadResult(result);
    },
  );

  server.tool(
    "find_component",
    "Search components on the loaded PCB (board, not schematic). `query` is a case-insensitive substring matched across reference, value AND footprint-id; the targeted filters narrow further. All supplied criteria combine with AND. Returns position and properties.",
    {
      query: z
        .string()
        .optional()
        .describe("Free-text substring matched across reference, value and footprint-id"),
      reference: z.string().optional().describe("Reference designator substring to match"),
      value: z.string().optional().describe("Component value substring to match"),
      footprint: z.string().optional().describe("Footprint-id substring to match"),
    },
    async ({ query, reference, value, footprint }) => {
      logger.debug(`Finding component (query=${query ?? ""} ref=${reference ?? ""})`);
      const result = await callKicadScript("find_component", {
        query,
        reference,
        value,
        footprint,
      });

      return formatKicadResult(result);
    },
  );

  server.tool(
    "get_component_properties",
    "Return all properties of a PCB component (position, rotation, layer, value, footprint).",
    {
      reference: z.string().describe("Reference designator of the component (e.g., 'R5')"),
    },
    async ({ reference }) => {
      logger.debug(`Getting properties for component: ${reference}`);
      const result = await callKicadScript("get_component_properties", {
        reference,
      });

      return formatKicadResult(result);
    },
  );

  server.tool(
    "add_component_annotation",
    "Add a text annotation/comment near a PCB component: places a PCB_TEXT at the component's position (plus optional offset) on a silkscreen or comments layer. Use offset to sit the label beside the part instead of on top of it.",
    {
      reference: z
        .string()
        .describe("Reference designator of the component to annotate (e.g., 'R5')"),
      text: z.string().describe("Annotation / comment text"),
      layer: z
        .string()
        .optional()
        .describe(
          "Target layer (default 'F.Silkscreen'); e.g. 'B.Silkscreen', 'User.Comments', 'F.Fab'. Legacy short names ('F.SilkS') are also accepted.",
        ),
      offset: z
        .object({
          x: z.number(),
          y: z.number(),
          unit: z.enum(["mm", "inch", "mil"]).optional(),
        })
        .optional()
        .describe("Offset from the component origin (default {0,0} mm — text lands at the part)"),
      size: z.number().optional().describe("Text height in mm (default 1.0)"),
    },
    passthrough("add_component_annotation"),
  );

  server.tool(
    "group_components",
    "Group PCB components into a named PCB_GROUP for easier selection. Refuses (creating nothing) if any reference is unknown. A component already in another group is moved into the new one; a group left empty by that move is removed — both reported.",
    {
      references: z
        .array(z.string())
        .describe("Reference designators to group (e.g., ['R1','R2'])"),
      groupName: z.string().describe("Name for the new group"),
    },
    passthrough("group_components"),
  );

  // ------------------------------------------------------
  // Replace Component Tool (DESTRUCTIVE)
  // ------------------------------------------------------
  server.tool(
    "replace_component",
    "Swap a placed footprint for a different library footprint. DESTRUCTIVE: deletes the old part and adds the new one, preserving reference, position, rotation and board side, and transferring pad nets by pad number. Pads that don't match on either side are reported.",
    {
      reference: z
        .string()
        .describe("Reference designator of the component to replace (e.g., 'U1')"),
      newFootprint: z
        .string()
        .describe("New footprint library id ('Library:Footprint' or bare 'Footprint')"),
      newValue: z
        .string()
        .optional()
        .describe("Optional new component value (defaults to the old value)"),
    },
    passthrough("replace_component"),
  );

  server.tool(
    "get_component_pads",
    "Return all pads of a component with exact positions, nets and sizes — use before routing; pass pad for just one.",
    {
      reference: z.string().describe("Reference designator of the component (e.g., 'U1')"),
      pad: z
        .string()
        .optional()
        .describe("Return only this pad number/name (e.g. '1', 'A1') instead of all pads"),
      unit: z.enum(["mm", "mil", "inch"]).optional().describe("Unit for coordinates (default: mm)"),
    },
    async ({ reference, pad, unit }) => {
      logger.debug(`Getting pads for component: ${reference}`);
      if (pad !== undefined) {
        return formatKicadResult(
          await callKicadScript("get_pad_position", { reference, pad, unit: unit || "mm" }),
        );
      }
      const result = await callKicadScript("get_component_pads", {
        reference,
        unit: unit || "mm",
      });

      return formatKicadResult(result);
    },
  );

  // ------------------------------------------------------
  // Edit Component Pad Tool
  //
  // Repairs pads on a PLACED footprint — broken library footprints
  // (e.g. easyeda BAT-SMD_CR1220-2: thru-hole pads with EMPTY numbers
  // and copper diameter == drill diameter) are otherwise unfixable:
  // edit_footprint_pad only patches .kicad_mod files and only numbered
  // pads. Targets by number OR zero-based index; refuses zero/negative
  // annular ring unless forced.
  // ------------------------------------------------------
  const padSizeSchema = z.union([
    z.number().describe("Uniform size (round/square)"),
    z.object({ x: z.number(), y: z.number() }),
    z.object({ w: z.number(), h: z.number() }),
  ]);
  server.tool(
    "edit_component_pad",
    "Edit pads of a PLACED footprint on the PCB (fix broken library footprints): copper size, drill, shape, pad number, pad type. Target by padNumber ('' matches unnumbered pads) or zero-based padIndex; several matches need all=true. Refuses a resulting annular ring <= 0 (copper<=drill) unless force=true. For .kicad_mod library files use edit_footprint_pad.",
    {
      reference: z.string().describe("Reference designator of the placed component (e.g. 'BT1')"),
      padNumber: z
        .union([z.string(), z.number()])
        .optional()
        .describe("Pad number to match; pass '' to target pads with empty numbers"),
      padIndex: z
        .number()
        .int()
        .optional()
        .describe("Zero-based pad index in get_component_pads order (works for unnumbered pads)"),
      padType: z
        .enum(["smd", "through_hole", "thru_hole", "npth", "connector"])
        .optional()
        .describe("Filter matches by pad type (or sole selector)"),
      size: padSizeSchema.optional().describe("New copper size in mm (number, or {x,y} / {w,h})"),
      drill: padSizeSchema
        .optional()
        .describe("New drill in mm (number = round; {x,y} or {w,h} = oval)"),
      shape: z.enum(["circle", "rect", "oval", "roundrect"]).optional().describe("New pad shape"),
      newPadNumber: z
        .union([z.string(), z.number()])
        .optional()
        .describe("Assign a pad number (fixes empty-number pads)"),
      newPadType: z
        .enum(["smd", "thru_hole", "npth", "connector"])
        .optional()
        .describe("Convert pad type"),
      all: z.boolean().optional().describe("Edit every matching pad (default: refuse multi-match)"),
      force: z.boolean().optional().describe("Allow zero/negative annular ring result"),
      unit: z.enum(["mm", "mil", "inch"]).optional().describe("Unit for sizes (default mm)"),
    },
    passthrough("edit_component_pad"),
  );

  server.tool(
    "get_component_list",
    "Return a list of all components on the PCB, optionally filtered by layer or bounding box region.",
    {
      layer: z.string().optional().describe("Filter by layer (e.g., 'F.Cu', 'B.Cu')"),
      boundingBox: boundingBoxFilter,
      unit: z.enum(["mm", "mil", "inch"]).optional().describe("Unit for coordinates (default: mm)"),
      ...paginationParams,
    },
    async ({ layer, boundingBox, unit, limit, offset }) => {
      logger.debug("Getting component list");
      const result = await callKicadScript("get_component_list", {
        layer,
        boundingBox,
        unit: unit || "mm",
        limit,
        offset,
      });

      return formatKicadResult(result);
    },
  );

  server.tool(
    "place_component_array",
    "Place a rectangular grid array of identical components on the PCB with configurable row/column spacing.",
    {
      componentId: z.string().describe("Component identifier"),
      startPosition: z
        .object({
          x: z.number(),
          y: z.number(),
          unit: z.enum(["mm", "inch", "mil"]).optional().describe("Unit (default mm)"),
        })
        .describe("Starting (top-left) position"),
      rows: z.number().describe("Number of rows"),
      columns: z.number().describe("Number of columns"),
      rowSpacing: z.number().describe("Spacing between rows (mm)"),
      columnSpacing: z.number().describe("Spacing between columns (mm)"),
      startReference: z
        .string()
        .optional()
        .describe("Starting reference; a trailing number is the first index (e.g. 'C5' → C5,C6,…)"),
      footprint: z.string().optional().describe("Footprint name"),
      value: z.string().optional().describe("Component value"),
      rotation: z.number().optional().describe("Rotation in degrees"),
    },
    async ({
      componentId,
      startPosition,
      rows,
      columns,
      rowSpacing,
      columnSpacing,
      startReference,
      footprint,
      value,
      rotation,
    }) => {
      logger.debug(`Placing component array: ${rows}x${columns} of ${componentId}`);
      const result = await callKicadScript("place_component_array", {
        componentId,
        startPosition,
        rows,
        columns,
        rowSpacing,
        columnSpacing,
        startReference,
        footprint,
        value,
        rotation,
      });

      return formatKicadResult(result);
    },
  );

  server.tool(
    "align_components",
    "Align multiple PCB components onto a common line: 'horizontal' shares one Y, 'vertical' shares one X, 'edge' snaps to a board edge. With `spacing` the parts are also evenly spaced (no overlap); with `referenceComponent` that part stays fixed and both the shared axis and the spacing sequence are anchored to it.",
    {
      references: z.array(z.string()).describe("Array of component references to align"),
      alignmentType: z
        .enum(["horizontal", "vertical", "edge"])
        .describe("Type of alignment: horizontal (shared Y), vertical (shared X), or edge"),
      spacing: z
        .number()
        .optional()
        .describe("Even spacing between adjacent components in mm (prevents overlap)"),
      referenceComponent: z
        .string()
        .optional()
        .describe("Anchor component: it stays fixed; the shared axis and spacing start from it"),
      edge: z
        .enum(["left", "right", "top", "bottom"])
        .optional()
        .describe("Board edge to align to (required when alignmentType is 'edge')"),
    },
    async ({ references, alignmentType, spacing, referenceComponent, edge }) => {
      logger.debug(`Aligning components: ${references.join(", ")}`);
      const result = await callKicadScript("align_components", {
        references,
        alignmentType,
        spacing,
        referenceComponent,
        edge,
      });

      return formatKicadResult(result);
    },
  );

  // ------------------------------------------------------
  // Check Courtyard Overlaps Tool
  //
  // Lets the caller validate a placement plan before committing it. The
  // `positions` parameter accepts hypothetical {ref: [x, y]} or
  // [x, y, rotation_degrees] entries; the board file is not modified.
  //
  // Approach ported from morningfire-pcb-automation
  //   https://github.com/NiNjA-CodE/morningfire-pcb-automation
  //   (scripts/placement/check_overlaps.py)
  // ------------------------------------------------------
  server.tool(
    "check_courtyard_overlaps",
    "Detect courtyard overlaps between footprints (read-only). Pass `positions` to validate hypothetical placements before committing. Returns overlap pairs with intersection extents (mm) and per-component boundary violations.",
    {
      positions: z
        .record(z.string(), z.array(z.number()).min(2).max(3))
        .optional()
        .describe(
          "Map ref -> [x, y] or [x, y, rotation_deg] in mm; listed refs are checked AS IF there, others use their board position",
        ),
      refs: z
        .array(z.string())
        .optional()
        .describe("Limit the check to these refs (default: every footprint on the board)."),
      margin: z
        .number()
        .optional()
        .describe(
          "Extra clearance in mm around every courtyard (default 0), e.g. a manufacturing keepout",
        ),
      include_boundary: z
        .boolean()
        .optional()
        .describe("Also flag courtyards past the board outline (default true)."),
      board_outline: z
        .object({
          x1: z.number(),
          y1: z.number(),
          x2: z.number(),
          y2: z.number(),
          unit: z.enum(["mm", "inch"]).optional(),
        })
        .optional()
        .describe("Board outline bbox override (default: derived from Edge.Cuts)."),
    },
    async (args) => {
      logger.debug(
        `Checking courtyard overlaps (virtual=${
          args.positions ? Object.keys(args.positions).length : 0
        })`,
      );
      const result = await callKicadScript("check_courtyard_overlaps", args);
      return formatKicadResult(result);
    },
  );

  server.tool(
    "duplicate_component",
    "Duplicate an existing PCB component at an offset position, optionally with a new reference designator.",
    {
      reference: z.string().describe("Reference of component to duplicate"),
      offset: z
        .object({
          x: z.number(),
          y: z.number(),
          unit: z.enum(["mm", "inch", "mil"]).optional(),
        })
        .describe("Offset from original position"),
      newReference: z.string().optional().describe("New reference designator"),
      count: z.number().optional().describe("Number of duplicates (default: 1)"),
    },
    async ({ reference, offset, newReference, count }) => {
      logger.debug(`Duplicating component: ${reference}`);
      const result = await callKicadScript("duplicate_component", {
        reference,
        offset,
        newReference,
        count,
      });

      return formatKicadResult(result);
    },
  );

  logger.info("Component management tools registered");
  server.tool(
    "auto_place_components",
    "Auto-place components with a connectivity-driven greedy heuristic: connected parts cluster, decoupling caps " +
      "hug their IC, courtyards keep spacing, positions snap to grid; power nets ignored for affinity. Returns HPWL " +
      "wirelength stats. A starting point, not a finished layout — review with get_board_2d_view and refine.",
    {
      components: z
        .array(z.string())
        .optional()
        .describe(
          "References to place (default: every unlocked footprint); others stay fixed and attract",
        ),
      fixedRefs: z
        .array(z.string())
        .optional()
        .describe("References to hold in place (they still exert affinity)"),
      spacing: z
        .number()
        .optional()
        .describe("Minimum courtyard-to-courtyard gap in mm (default 1.0)"),
      grid: z.number().optional().describe("Placement grid in mm (default 0.5)"),
      area: z
        .object({ x1: z.number(), y1: z.number(), x2: z.number(), y2: z.number() })
        .optional()
        .describe("Placement area in mm (default: the board outline)"),
      dryRun: z
        .boolean()
        .optional()
        .describe("Compute placements without moving footprints (default false)"),
      includeMechanical: z
        .boolean()
        .optional()
        .describe(
          "Also relocate netless mechanical footprints (mounting holes H/MH, fiducials FID, test points TP). " +
            "Default false: they stay fixed and are reported in skipped_mechanical",
        ),
    },
    passthrough("auto_place_components"),
  );
}
